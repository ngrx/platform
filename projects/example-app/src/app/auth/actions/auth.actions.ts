import { createAction } from '@ngrx/store';

export const logout = createAction('[Auth] Logout');
export const logoutConfirmation = createAction('[Auth] Logout Confirmation');
export const logoutConfirmationDismiss = createAction(
  '[Auth] Logout Confirmation Dismiss'
);
