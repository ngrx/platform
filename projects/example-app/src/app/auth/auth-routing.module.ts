import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { LoginPageComponent } from './containers/login-page.component';

const routes: Routes = [{ path: 'login', component: LoginPageComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class AuthRoutingModule {}
