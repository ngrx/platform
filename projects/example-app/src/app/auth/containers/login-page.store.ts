import { Injectable } from '@angular/core';
import {
  ComponentStore,
  OnStoreInit,
  OnStateInit,
} from '@ngrx/component-store';
import { tap } from 'rxjs';

interface LifeCycle {
  init: boolean;
}

@Injectable()
export class LoginPageStore
  extends ComponentStore<LifeCycle>
  implements OnStoreInit, OnStateInit
{
  constructor() // private readonly service: Service
  {
    super({ init: true });
  }

  logEffect = this.effect(tap<string>(console.log));

  ngrxOnStoreInit() {
    console.log('onInitStore');
    // console.log('service', this.service); // undefined
    console.log('log effect', this.logEffect('one')); // undefined
    // console.log('effect')
  }

  ngrxOnStateInit() {
    console.log('onInitState');
    // console.log('service', this.service); // undefined
    console.log('log effect', this.logEffect('two')); // undefined
  }
}
