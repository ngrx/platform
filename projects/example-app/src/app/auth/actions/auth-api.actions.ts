import { props, createAction } from '@ngrx/store';
import { User } from '@example-app/auth/models';

export const loginSuccess = createAction(
  '[Auth/API] Login Success',
  props<{ user: User }>()
);

export const loginFailure = createAction(
  '[Auth/API] Login Failure',
  props<{ error: any }>()
);

export const loginRedirect = createAction('[Auth/API] Login Redirect');
