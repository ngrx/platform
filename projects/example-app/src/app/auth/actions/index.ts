import { AuthActions, AuthActionTypes } from './auth.actions';
import { AuthApiActions, AuthApiActionTypes } from './auth-api.actions';
import { LoginPageActions, LoginPageActionTypes } from './login-page.actions';

export {
  AuthActions,
  AuthActionTypes,
  AuthApiActions,
  AuthApiActionTypes,
  LoginPageActions,
  LoginPageActionTypes,
};
