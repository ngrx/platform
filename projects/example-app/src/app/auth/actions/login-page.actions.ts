import { createAction, props } from '@ngrx/store';
import { Credentials } from '@example-app/auth/models';

export const login = createAction(
  '[Login Page] Login',
  props<{ credentials: Credentials }>()
);
