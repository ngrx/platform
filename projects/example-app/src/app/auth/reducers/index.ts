import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers,
} from '@ngrx/store';
import * as fromRoot from '@example-app/reducers';
import * as fromAuth from '@example-app/auth/reducers/auth.reducer';
import * as fromLoginPage from '@example-app/auth/reducers/login-page.reducer';

export const authFeatureKey = 'auth';

export interface AuthState {
  [fromAuth.statusFeatureKey]: fromAuth.State;
  [fromLoginPage.loginPageFeatureKey]: fromLoginPage.State;
}

export interface State extends fromRoot.State {
  [authFeatureKey]: AuthState;
}

export function reducers(state: AuthState | undefined, action: Action) {
  return combineReducers({
    [fromAuth.statusFeatureKey]: fromAuth.reducer,
    [fromLoginPage.loginPageFeatureKey]: fromLoginPage.reducer,
  })(state, action);
}

export const selectAuthState = createFeatureSelector<State, AuthState>(
  authFeatureKey
);

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state) => state.status
);
export const selectUser = createSelector(
  selectAuthStatusState,
  fromAuth.getUser
);
export const selectLoggedIn = createSelector(selectUser, (user) => !!user);

export const selectLoginPageState = createSelector(
  selectAuthState,
  (state) => state.loginPage
);
export const selectLoginPageError = createSelector(
  selectLoginPageState,
  fromLoginPage.getError
);
export const selectLoginPagePending = createSelector(
  selectLoginPageState,
  fromLoginPage.getPending
);
