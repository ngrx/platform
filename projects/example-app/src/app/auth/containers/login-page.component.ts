import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { Credentials } from '@example-app/auth/models';
import * as fromAuth from '@example-app/auth/reducers';
import { LoginPageActions } from '@example-app/auth/actions';

@Component({
  selector: 'bc-login-page',
  template: `
    <bc-login-form
      (submitted)="onSubmit($event)"
      [pending]="pending$ | async"
      [errorMessage]="error$ | async"
    >
    </bc-login-form>
  `,
  styles: [],
})
export class LoginPageComponent {
  pending$ = this.store.select(fromAuth.selectLoginPagePending);
  error$ = this.store.select(fromAuth.selectLoginPageError);

  constructor(private store: Store) {}

  onSubmit(credentials: Credentials) {
    this.store.dispatch(LoginPageActions.login({ credentials }));
  }
}
