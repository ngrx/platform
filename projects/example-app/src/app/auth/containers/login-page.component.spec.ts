import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { LoginPageComponent } from '@example-app/auth/containers';
import { LoginFormComponent } from '@example-app/auth/components';
import * as fromAuth from '@example-app/auth/reducers';
import { LoginPageActions } from '@example-app/auth/actions';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { MaterialModule } from '@example-app/material';

describe('Login Page', () => {
  let fixture: ComponentFixture<LoginPageComponent>;
  let store: MockStore;
  let instance: LoginPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, MaterialModule, ReactiveFormsModule],
      declarations: [LoginPageComponent, LoginFormComponent],
      providers: [
        provideMockStore({
          selectors: [
            { selector: fromAuth.selectLoginPagePending, value: false },
          ],
        }),
      ],
    });

    fixture = TestBed.createComponent(LoginPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.inject(MockStore);

    spyOn(store, 'dispatch');
  });

  /**
   * Container components are used as integration points for connecting
   * the store to presentational components and dispatching
   * actions to the store.
   *
   * Container methods that dispatch events are like a component's output observables.
   * Container properties that select state from store are like a component's input properties.
   * If pure components are functions of their inputs, containers are functions of state
   *
   * Traditionally you would query the components rendered template
   * to validate its state. Since the components are analogous to
   * pure functions, we take snapshots of these components for a given state
   * to validate the rendered output and verify the component's output
   * against changes in state.
   */
  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a login event on submit', () => {
    const credentials: any = {};
    const action = LoginPageActions.login({ credentials });

    instance.onSubmit(credentials);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
