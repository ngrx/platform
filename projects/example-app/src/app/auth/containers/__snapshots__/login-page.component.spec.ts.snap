// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should compile 1`] = `
<bc-login-page
  __ngContext__={[Function LRootView]}
  error$={[Function Store]}
  pending$={[Function Store]}
  store={[Function MockStore]}
>
  <bc-login-form>
    <mat-card
      class="mat-card mat-focus-indicator _mat-animation-noopable"
    >
      <mat-card-title
        class="mat-card-title"
      >
        Login
      </mat-card-title>
      <mat-card-content
        class="mat-card-content"
      >
        <form
          class="ng-untouched ng-pristine ng-valid"
          novalidate=""
        >
          <p>
            <mat-form-field
              class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-should-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
            >
              <div
                class="mat-form-field-wrapper"
              >
                <div
                  class="mat-form-field-flex"
                >
                  <div
                    class="mat-form-field-infix"
                  >
                    <input
                      aria-invalid="false"
                      aria-required="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                      data-placeholder="Username"
                      formcontrolname="username"
                      matinput=""
                      type="text"
                    />
                    <span
                      class="mat-form-field-label-wrapper"
                    >
                      <label
                        class="mat-form-field-label"
                      >
                        <span>
                          Username
                        </span>
                      </label>
                    </span>
                  </div>
                </div>
                <div
                  class="mat-form-field-underline"
                >
                  <span
                    class="mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-form-field-subscript-wrapper"
                >
                  <div
                    class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                    style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                  >
                    <div
                      class="mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
          </p>
          <p>
            <mat-form-field
              class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
            >
              <div
                class="mat-form-field-wrapper"
              >
                <div
                  class="mat-form-field-flex"
                >
                  <div
                    class="mat-form-field-infix"
                  >
                    <input
                      aria-invalid="false"
                      aria-required="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                      data-placeholder="Password"
                      formcontrolname="password"
                      matinput=""
                      type="password"
                    />
                    <span
                      class="mat-form-field-label-wrapper"
                    >
                      <label
                        class="mat-form-field-label mat-empty mat-form-field-empty"
                      >
                        <span>
                          Password
                        </span>
                      </label>
                    </span>
                  </div>
                </div>
                <div
                  class="mat-form-field-underline"
                >
                  <span
                    class="mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-form-field-subscript-wrapper"
                >
                  <div
                    class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                    style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                  >
                    <div
                      class="mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
          </p>
          <p
            class="loginButtons"
          >
            <button
              class="mat-focus-indicator mat-button mat-button-base _mat-animation-noopable"
              mat-button=""
              type="submit"
            >
              <span
                class="mat-button-wrapper"
              >
                Login
              </span>
              <span
                class="mat-ripple mat-button-ripple"
                matripple=""
              />
              <span
                class="mat-button-focus-overlay"
              />
            </button>
          </p>
        </form>
      </mat-card-content>
    </mat-card>
  </bc-login-form>
</bc-login-page>
`;
